gpjax/decision_making/posterior_optimizer.py

```diff
@@ -0,0 +1,104 @@
+# Copyright 2023 The JaxGaussianProcesses Contributors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+from abc import (
+    ABC,
+    abstractmethod,
+)
+from dataclasses import dataclass
+
+import optax as ox
+
+import gpjax as gpx
+from gpjax.dataset import Dataset
+from gpjax.gps import AbstractPosterior
+from gpjax.objectives import AbstractObjective
+from gpjax.typing import KeyArray
+
+
+@dataclass
+class AbstractPosteriorOptimizer(ABC):
+    """
+    Abstract base class for optimizing a GP posterior.
+    """
+
+    @abstractmethod
+    def optimize(
+        self,
+        posterior: AbstractPosterior,
+        dataset: Dataset,
+        key: KeyArray,
+    ) -> AbstractPosterior:
+        """
+        Take a posterior and corresponding dataset and optimizes the posterior
+        hyperparameters.
+
+        Args:
+            posterior: Posterior being optimized.
+            dataset: Dataset used for optimizing posterior.
+            key: A JAX PRNG key for generating random numbers.
+        """
+        raise NotImplementedError
+
+
+@dataclass
+class AdamPosteriorOptimizer(AbstractPosteriorOptimizer):
```

Comment:
This feels a level of abstraction too far. What if I want to use `adamw` or `sgd`? With this abstraction I'd have subclass `AbstractPosteriorOptimizer` just to change the optimiser. Could we instead structure it as `OptaxPosteriorOptimizer` or something similar where the user supplies the relevant optimiser? It feels a bit more general.

Reply:
Further, what's the benefit of this object, given that it just calls `fit`?

Reply:
I agree with @thomaspinder. I think you just need to pass in optim and num_iters into the PosteriorHandler

Reply:
Yes - I had originally kept the optimisation logic separate in anticipation of moving towards fitting models with L-BFGS further down the line. However, I think I will move this logic to the `PosteriorHandler` class, and accept that I may just have to change this class a bit once I switch to using L-BFGS.